### PR TITLE
Use ConcurrentDictionary for resolver cache

### PR DESCRIPTION
**Pull request by joelverhagen**
Address https://github.com/NuGet/Home/issues/2646.

@emgarten @yishaigalatzer @pranavkm

### CODE REVIEW

File: src/NuGet.Core/NuGet.ProjectModel/PackageSpecReferenceDependencyProvider.cs [hunk 1]
```diff
@@ -323,16 +325,9 @@ private IPackageSpecResolver GetPackageSpecResolver(string rootPath)
...
-                if (_resolverCache.TryGetValue(rootPath, out cachedResolver))
-                {
-                    specResolver = cachedResolver;
-                }
-                else
-                {
-                    specResolver = new PackageSpecResolver(rootPath);
-                    _resolverCache.Add(rootPath, specResolver);
-                }
+                specResolver = _resolverCache.GetOrAdd(
```
Comment (pranavkm) [inline src/NuGet.Core/NuGet.ProjectModel/PackageSpecReferenceDependencyProvider.cs:328]: The old code pattern wouldn't have allocated a delegate on a per call basis. Maybe leaving it as is is better

**Comment by pranavkm** [comment]
:shipit:

**Comment by yishaigalatzer** [comment]
How did we validate the fix?

**Comment by yishaigalatzer** [comment]
We should merge quickly into the CLI no later than Monday

**Comment by joelverhagen** [comment]
Ran some simulations to verify that adding to a dictionary in a `ConcurrentDictionary` factory (which is essentially happening in the code) can cause a problem. This is pretty obvious because `ConcurrentDictionary` does not guarantee a single factory execution per key. I have also ran our automated tests.

File: src/NuGet.Core/NuGet.ProjectModel/PackageSpecReferenceDependencyProvider.cs (hunk 1 above)
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.ProjectModel/PackageSpecReferenceDependencyProvider.cs:328]: Addressed. Thanks!

**Comment by emgarten** [comment]
:shipit:

**Pull request merged**